erpnext/stock/doctype/delivery_trip/delivery_trip.py

```diff
@@ -28,14 +30,76 @@ def validate(self):
 		self.validate_stop_addresses()
 		self.update_status()
 		self.update_package_total()
+		self.generate_directions_url()
+		self.link_invoice_against_trip()
+
+	def generate_directions_url(self):
+		if not frappe.db.get_single_value("Google Settings", "enable"):
+			return
+
+		if not self.driver_address:
+			return
+
+		if self.delivery_stops:
+			route_list = self.form_route_list(optimize=False)
+
+			if not route_list:
+				return
+
+			route_list = route_list[0]
+
+			context = {
+				"key": get_decrypted_password("Google Settings", "Google Settings", "api_key"),
+				"origin": quote(route_list[0], safe=''),
+				"destination": quote(route_list[-1], safe=''),
+				"waypoints": quote('|'.join(route_list[1:-1]), safe='')
+			}
+
+			map_data = frappe.render_template("templates/includes/google_maps_embed.html", context)
+			self.map_embed = map_data
+
+	def link_invoice_against_trip(self):
+		for delivery_stop in self.delivery_stops:
+			if delivery_stop.delivery_note:
+				sales_invoice = frappe.get_all("Delivery Note Item",
+					filters={"docstatus": 1, "parent": delivery_stop.delivery_note},
+					fields=["against_sales_invoice"],
+					distinct=True)
+
+				if sales_invoice and len(sales_invoice) == 1:
+					delivery_stop.sales_invoice = sales_invoice[0].against_sales_invoice
+				# set sales invoice for cyecle sales order>>Delivery Note >>sales invoice>> then Delivery Trip from Delivery Note
+				if sales_invoice[0].against_sales_invoice==None:
+					sales_invoice = frappe.get_all("Sales Invoice Item",
+					filters={"docstatus": 1, "delivery_note": delivery_stop.delivery_note},
+					fields=["distinct(parent)"])
+					if len(sales_invoice)==1:
+						delivery_stop.sales_invoice = sales_invoice[0].parent
 
 	def on_submit(self):
 		self.update_delivery_notes()
+		self.make_transfer_templates()
+
+	def make_transfer_templates(self):
+		for stop in self.delivery_stops:
+			if not stop.delivery_note:
+				continue
+
+			for item in frappe.get_doc("Delivery Note", stop.delivery_note).items:
+				if frappe.db.get_value("Item", item.item_code, "is_compliance_item"):
+					make_integration_request("Delivery Note", stop.delivery_note, "Transfer")
+					break
```

Comment:
put all of the doc-events first, then add these functions after the doc events for proper readability